draft/0002-subdocapi.md

```diff
@@ -0,0 +1,380 @@
+
+* Throws `PathExists` if the unique element already exists
+* Throws `CantInsertValue` if the value is not a JSON primitive.
+* Throws `PathMismatch` if the array already contains non-primitive elements
+
+> Because there is too much variation in semantics between `add_unique_in` and
+> `extend_in`, they are separate methods rather than a single unified one.
+
+### `counter_in(docid, path, delta, create_parents, cas, ttl, persist_to, replicate_to)`
+Performs an atomic counter operation on a given path.
+
+* `delta`: The numeric delta, this is simply added (`+`) to the existing value.
+  This value may be signed to 'decrement' or unsigned to 'increment' the existing
+  value.
+
+See "Numeric Errors" below for a list of errors thrown by this operation.
+This method also returns the current value as a `DocumentFragment`
+
+### `remove_in(docid, path, cas, ttl, persist_to, replicate_to)`
+Removes a path from the document
+
+## Multi-Path methods
+
+These methods take a document ID and a list of _specifications_ to perform. There are
+only two methods (`mutate_in` and `lookup_in`).
+
+The specification of each method is highly language specific, to use pseudo code, we will
+pretend there are two static helper classes which generate appropriate specs:
+
+### `LookupSpec`
+Creates lookup specifications. There are two 'static methods'
+
+* `LookupSpec.get(docid)` (takes the same arguments as `get_in`
+* `LookupSpec.exists(docid)` (takes the same arguments as `exists_in`)
+
+Both these objects should return a spec suitable for consumption by the `lookup_in`
+method.
+
+### `MutationSpec`
+Creates mutation specifications. There are several 'static methods' which may be used.
+These all function exactly like their single-path counterparts except that the document
+ID and cas are not provided (these are passed directly into `mutate_in`).
+
+* `MutationSpec.replace(path, value)`: See `replace_in`
+* `MutationSpec.upsert(path, value, create_parents)`. See `upsert_in`
+* `MutationSpec.insert(path, value, create_parents)`. See `insert_in`
+* `MutationSpec.extend(path, value_or_values, direction, create_parents)`. See `extend_in`
+* `MutationSpec.arrayinsert(path, value_or_values)`. See `arrayinsert_in`
+* `MutationSpec.addunique(path, value)`. See `addunique_in`
+* `MutationSpec.counter(path, delta)`. See `counter_in`
+* `MutationSpec.remove(path)`. See `remove_in`
+
+### `LookupResultSpec`
+This object represents a result for a single path when using `lookup_in`.
+It contains two fields:
+
+* `status`: The error code for the operation. If the path exists, this should
+  contain an object or value which indicates "Success" (this can be a NULL or
+  empty value to indicate no error)
+* `value`: The value for the path. This is always empty when using `exists`.
+
+
+### `lookup_in(docid, lookup_specs...)`
+
+Retrieve (or check the existence of) multiple paths.
+
+Returns a list of MetaDocFragment objects. Individual objects may also
```

Comment:
if we dont put the path into the fragment, wouldn't then then actually be one DocumentFragment with a list of actual fragments inside? Of course if we put the path into the fragment its ambiguous...

I'm asking since most of the stuff is duplicated (like the id(), cas().. and so on)..